### DISCUSSION TITLE
Optimizing textcat

### DISCUSSION THREAD

**Discussion by rvandomme** [explosion/spaCy#7637] [Help: Coding & Implementations] [answered]
Hello spaCy community!
Looking for some help.

I am working on a text categorization project.
I have 120,000 documents, ranging from 300 to 30,000 characters.
Documents are professional, banking, finance, legal, HR, etc
They are assigned to 11 different classes, one being a "Other". 
Classes are disbalanced ranging from 30000 to 1000.
The text is clean, symbols, punctuation, numbers and stop words are removed.

A colleague used fasttext and got 77.3% accuracy in a few minutes.
I am a big fan of spaCy and determined to do better.
Unfortunately, I struggle with version 3 and can barely pass 30% accuracy. :-(
My config is below.

I work on AzureML environment.
Any hint, advise or suggestion to make it better would be very appreciated.
Thanks in advance
Roger

Here is the config I use:

```
[paths]
train = "./data/train-FTS.spacy"
dev = "./data/valid-FTS.spacy"
vectors = null
init_tok2vec = null

[system]
gpu_allocator = null
seed = 0

[nlp]
lang = "en"
pipeline = ["tok2vec","textcat"]
batch_size = 1000
disabled = []
before_creation = null
after_creation = null
after_pipeline_creation = null
tokenizer = {"@tokenizers":"spacy.Tokenizer.v1"}

[components]

[components.textcat]
factory = "textcat"
threshold = 0.5

[components.textcat.model]
@architectures = "spacy.TextCatEnsemble.v2"
nO = null

[components.textcat.model.linear_model]
@architectures = "spacy.TextCatBOW.v1"
exclusive_classes = true
ngram_size = 1
no_output_layer = false
nO = null

[components.textcat.model.tok2vec]
@architectures = "spacy.Tok2VecListener.v1"
width = ${components.tok2vec.model.encode.width}
upstream = "*"

[components.tok2vec]
factory = "tok2vec"

[components.tok2vec.model]
@architectures = "spacy.Tok2Vec.v2"

[components.tok2vec.model.embed]
@architectures = "spacy.MultiHashEmbed.v1"
width = ${components.tok2vec.model.encode.width}
attrs = ["ORTH","SHAPE"]
rows = [5000,2500]
include_static_vectors = true

[components.tok2vec.model.encode]
@architectures = "spacy.MaxoutWindowEncoder.v2"
width = 256
depth = 8
window_size = 1
maxout_pieces = 3

[corpora]

[corpora.dev]
@readers = "spacy.Corpus.v1"
path = ${paths.dev}
max_length = 0
gold_preproc = false
limit = 0
augmenter = null

[corpora.train]
@readers = "spacy.Corpus.v1"
path = ${paths.train}
max_length = 2000
gold_preproc = false
limit = 0
augmenter = null

[training]
dev_corpus = "corpora.dev"
train_corpus = "corpora.train"
seed = ${system.seed}
gpu_allocator = ${system.gpu_allocator}
dropout = 0.1
accumulate_gradient = 1
patience = 1600
max_epochs = 0
max_steps = 20000
eval_frequency = 200
frozen_components = []
before_to_disk = null

[training.batcher]
@batchers = "spacy.batch_by_words.v1"
discard_oversize = false
tolerance = 0.2
get_length = null

[training.batcher.size]
@schedules = "compounding.v1"
start = 100
stop = 1000
compound = 1.001
t = 0.0

[training.logger]
@loggers = "spacy.ConsoleLogger.v1"
progress_bar = false

[training.optimizer]
@optimizers = "Adam.v1"
beta1 = 0.9
beta2 = 0.999
L2_is_weight_decay = true
L2 = 0.01
grad_clip = 1.0
use_averages = false
eps = 0.00000001
learn_rate = 0.001

[training.score_weights]
cats_score_desc = null
cats_micro_p = null
cats_micro_r = null
cats_micro_f = null
cats_macro_p = null
cats_macro_r = null
cats_macro_f = null
cats_macro_auc = null
cats_f_per_type = null
cats_macro_auc_per_type = null
cats_score = 1.0

[pretraining]

[initialize]
vectors = "en_core_web_trf"
init_tok2vec = ${paths.init_tok2vec}
vocab_data = null
lookups = null
before_init = null
after_init = null

[initialize.components]

[initialize.tokenizer]
```

**Comment by polm** [comment, marked as answer]
Sorry to hear you're having trouble with your model.

One thing that jumps out at me about your setup is this:

> The text is clean, symbols, punctuation, numbers and stop words are removed.

In general with modern NLP models you shouldn't modify the text like this, in particular stop word removal is a bad idea. For a BOW model it shouldn't make a difference, but I would suggest trying with text that looks like natural sentences - so if you have weird whitespace or decorative unicode or something you can clean that up, but if it looks like a normal newspaper article already that's great.

```
[initialize]
vectors = "en_core_web_trf"
```

This is a bit strange - the vectors argument here is for [static vectors](https://spacy.io/usage/embeddings-transformers#static-vectors), not Transformers, which are contextual. You should use static vectors like those distributed with the medium and large models.

What specifically happens in the code is that the `vectors` of the model are loaded, but in the case of Transformers models there are no `vectors` - Transformers representations are handled differently. I think what happens is this causes all your words to be treated as unks. Try replacing `en_core_web_trf` here with `en_core_web_lg`. 

Also, for the record, it looks like your config is based on the "efficient CPU" configuration, but you might want to try something optimized for accuracy if you have good enough hardware.
